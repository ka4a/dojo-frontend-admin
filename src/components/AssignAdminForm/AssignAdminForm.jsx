import React, { useEffect, useState, memo } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { Button, Col, Form } from '@edx/paragon';
import * as R from 'ramda';
import {
  assignStaffMember,
  fetchRoles,
} from '../../store/adminDashboard/adminDashboard.thunk';
import {
  getDashboardNewInstructor,
  getDashboardRoles,
} from '../../store/adminDashboard/adminDashboard.selector';
import { validateEmail, validateRole } from '../../utils/helpers/validation';

function AssignAdminForm({ courseId }) {
  const newInstructor = useSelector(getDashboardNewInstructor);
  const roles = useSelector(getDashboardRoles);

  const dispatch = useDispatch();

  const [emailFieldStatus, setEmailFieldStatus] = useState({
    isValid: true,
    errors: {},
  });
  const [roleFieldStatus, setRoleFieldStatus] = useState({
    isValid: true,
    errors: {},
  });

  const handleFormSubmit = (event) => {
    event.preventDefault();

    const email = event.target.email.value;
    const role = event.target.role.value;

    const emailStatus = validateEmail(email);
    const roleStatus = validateRole(role);

    setEmailFieldStatus(emailStatus);
    setRoleFieldStatus(roleStatus);

    if (!R.isEmpty(emailStatus.errors) || !R.isEmpty(roleStatus.errors)) {
      return;
    }

    dispatch(
      assignStaffMember({
        userEmail: email,
        courseId,
        role,
      }),
    );
  };

  useEffect(() => {
    dispatch(fetchRoles());
  }, [dispatch]);

  if (R.isEmpty(emailFieldStatus.errors) && newInstructor.error) {
    emailFieldStatus.errors = newInstructor.error;
  }
  const emailFieldIsInvalid = !R.isEmpty(emailFieldStatus.errors);
  const roleFieldIsInvalid = !R.isEmpty(roleFieldStatus.errors);

  return (
    <Form onSubmit={handleFormSubmit}>
      <Form.Row>
        <Form.Group
          as={Col}
          controlId="formGridEmail"
          isInvalid={emailFieldIsInvalid}
        >
          <Form.Control type="email" floatingLabel="Email" name="email" />
          {emailFieldIsInvalid && (
            <Form.Control.Feedback type="invalid">
              {emailFieldStatus.errors}
            </Form.Control.Feedback>
          )}
        </Form.Group>
        <Form.Group
          as={Col}
          controlId="formGridState"
          isInvalid={roleFieldIsInvalid}
        >
          <Form.Control floatingLabel="Role" as="select" name="role">
            <option value="">Select a role</option>
            {roles.status === 'resolved'
              && roles.data?.map((role) => (
                <option key={role} value={role}>
                  {role}
                </option>
              ))}
          </Form.Control>
          {roleFieldIsInvalid && (
            <Form.Control.Feedback type="invalid">
              {roleFieldStatus.errors}
            </Form.Control.Feedback>
          )}
        </Form.Group>
      </Form.Row>
      <Button variant="primary" type="submit" className="mt-3 mb-2">
        Assign
      </Button>
    </Form>
  );
}

AssignAdminForm.propTypes = {
  courseId: PropTypes.number.isRequired,
};

export default memo(AssignAdminForm);
