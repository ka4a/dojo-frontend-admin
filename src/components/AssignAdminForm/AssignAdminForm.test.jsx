import { shallow, mount } from 'enzyme';
import { AdminDashboardProvider } from '../../test/AdminDashboardProvider';
import AssignAdminForm from './AssignAdminForm';

const defaultProps = {
  courseId: 1,
};

describe('AssignAdminForm', () => {
  it('it should render AssignAdminForm component', () => {
    shallow(
      <AdminDashboardProvider>
        <AssignAdminForm {...defaultProps} />
      </AdminDashboardProvider>,
    );
  });

  it('it should render the expected HTML', () => {
    expect(
      mount(
        <AdminDashboardProvider>
          <AssignAdminForm {...defaultProps} />
        </AdminDashboardProvider>,
      ).html(),
    ).toMatchSnapshot();
  });
});
