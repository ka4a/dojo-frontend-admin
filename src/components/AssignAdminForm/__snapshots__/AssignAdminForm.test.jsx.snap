// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssignAdminForm it should render the expected HTML 1`] = `"<form class=\\"\\"><div class=\\"form-row\\"><div class=\\"pgn__form-group col\\"><div class=\\"pgn__form-control-decorator-group has-floating-label\\"><input name=\\"email\\" type=\\"email\\" id=\\"formGridEmail\\" class=\\"form-control\\"><div class=\\"pgn__form-control-floating-label\\"><div class=\\"pgn__form-control-floating-label-content\\"><label class=\\"pgn__form-control-floating-label-text\\" for=\\"formGridEmail\\">Email</label></div></div></div></div><div class=\\"pgn__form-group col\\"><div class=\\"pgn__form-control-decorator-group has-floating-label\\"><select name=\\"role\\" id=\\"formGridState\\" class=\\"form-control\\"><option value=\\"\\">Select a role</option><option value=\\"grader\\">grader</option><option value=\\"instructor\\">instructor</option></select><div class=\\"pgn__form-control-floating-label\\"><div class=\\"pgn__form-control-floating-label-content\\"><label class=\\"pgn__form-control-floating-label-text\\" for=\\"formGridState\\">Role</label></div></div></div></div></div><button type=\\"submit\\" class=\\"mt-3 mb-2 btn btn-primary\\">Assign</button></form>"`;
